src/app/core/services/node-data/provider/vmware-cloud-director.ts

```diff
@@ -0,0 +1,187 @@
+                .pipe(
+                  catchError(_ => {
+                    if (onError) {
+                      onError();
+                    }
+
+                    return onErrorResumeNext(of([]));
+                  })
+                )
+            )
+          );
+      case NodeDataMode.Dialog:
+        return this._projectService.selectedProject
+          .pipe(debounceTime(this._debounce))
+          .pipe(tap(_ => (onLoadingCb ? onLoadingCb() : null)))
+          .pipe(
+            switchMap(project =>
+              this._vmwareCloudDirectorService.getStorageProfiles(project.id, this._clusterSpecService.cluster.id)
+            )
+          )
+          .pipe(
+            catchError(_ => {
+              if (onError) {
+                onError();
+              }
+
+              return onErrorResumeNext(of([]));
+            })
+          )
+          .pipe(take(1));
+    }
+  }
+
+  catalogs(onError: () => void = undefined, onLoadingCb: () => void = null): Observable<VMwareCloudDirectorCatalog[]> {
+    switch (this._nodeDataService.mode) {
+      case NodeDataMode.Wizard:
+        return this._clusterSpecService.clusterChanges
+          .pipe(filter(_ => this._clusterSpecService.provider === NodeProvider.VMWARECLOUDDIRECTOR))
+          .pipe(debounceTime(this._debounce))
+          .pipe(map(() => this._clusterSpecService.cluster))
+          .pipe(
+            switchMap(cluster =>
+              this._getProvider(cluster)
+                .catalogs(cluster.spec.cloud.dc, onLoadingCb)
+                .pipe(
+                  catchError(_ => {
+                    if (onError) {
+                      onError();
+                    }
+
+                    return onErrorResumeNext(of([]));
+                  })
+                )
+            )
+          );
+      case NodeDataMode.Dialog:
+        return this._projectService.selectedProject
+          .pipe(debounceTime(this._debounce))
+          .pipe(tap(_ => (onLoadingCb ? onLoadingCb() : null)))
+          .pipe(
+            switchMap(project =>
+              this._vmwareCloudDirectorService.getCatalogs(project.id, this._clusterSpecService.cluster.id)
+            )
+          )
+          .pipe(
+            catchError(_ => {
+              if (onError) {
+                onError();
+              }
+
+              return onErrorResumeNext(of([]));
+            })
+          )
+          .pipe(take(1));
+    }
+  }
+
+  templates(
+    catalogName: string,
+    onError: () => void = undefined,
+    onLoadingCb: () => void = null
+  ): Observable<VMwareCloudDirectorTemplate[]> {
+    switch (this._nodeDataService.mode) {
+      case NodeDataMode.Wizard:
+        return this._getProvider(this._clusterSpecService.cluster)
+          .templates(this._clusterSpecService.cluster.spec.cloud.dc, catalogName, onLoadingCb)
+          .pipe(
+            catchError(_ => {
+              if (onError) {
+                onError();
```

Comment:
Why do we need this? I hate magic substitutions and would prefer using explicit if condition, just like we do everywhere else.

Reply:
I don't like the idea of using optional chaining for each and everything. @KhizerRehan @Waseem826 @ahmadhamzh(sig-ui) Opinions?

Reply:
IMO, it can be used within `template` but changing all over the `.ts` files 
- Make more code complex and inconsistent with all over the application
- It does make code shorter but makes difficult to understand by other devs
- I think tradeoff of writing an extra line instead of adding `?.` is better to for Readability point of view.

But i do agree can be used in `templates` or define getters in `ts` file to use that to  avoid `?.` from template file and use `?.` on getters

Reply:
Not making this change as agreed in conversation above.